feat: ecocredit ORM proto definitions

technicallyty opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

adds ORM proto definitions for x/ecocredit

Closes: n/a

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,213 @@
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 3,
+        primary_key: {fields: "batch_denom"}
```
Let's actually add an uint64 auto increment primary key and make this an unique index. I think we'll want a short uint64 to reference batches in the order book and denoms are getting kind of long

technicallyty replied:
are there docs anywhere for the syntax on this?

aaronc replied:
in orm.proto
```suggestion
        primary_key: {fields: "id", auto_increment: true}
        index: {fields: "batch_denom", unique: true}
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,213 @@
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "class_id"}
```
Maybe let's add admin and credit type indexes?

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/files#diff-302ae7137e2a8ea4a04301cf72f3df8311c99ea4716efaf59ae04291e030fb63R16-R17

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,213 @@
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "project_id"}
```
How about issuer and class indexes?

technicallyty replied:
added: https://github.com/regen-network/regen-ledger/pull/700/files#diff-302ae7137e2a8ea4a04301cf72f3df8311c99ea4716efaf59ae04291e030fb63R56-R57

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,213 @@
+}
+
+// SellOrder represents the information for a sell order.
+message SellOrder {
```
Let's move orders into a separate proto package

technicallyty replied:
i separated it like:

`ecocredit/v1beta1/marketplace/state.proto`


is that how we want it?

aaronc replied:
Yep 👍

aaronc reviewed (commented):
Great start @technicallyty ! Left a few suggestions

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,213 @@
+    string admin = 2;
+
+    // issuers are the approved issuers of the credit class.
+    repeated string issuers = 3;
```
In keeping with 1NF (which I think we should), let's move this to its own join table

aaronc commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,96 @@
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "order_id"}
+        index: {id: 1 fields: "batch_denom"}
```
how about owner index?

aaronc commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,96 @@
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "buy_order_id"}
```
buyer index too?

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,139 @@
+message ClassIssuers {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "class_id"}
```
```suggestion
        primary_key: {fields: "class_id,issuer"}
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,139 @@
+    string class_id = 1;
+
+    // issuers are the approved issuers of the credit class.
+    repeated string issuers = 2;
```
```suggestion
   string issuer = 2;
```

in 1NF we can't have any `repeated` fields

aaronc reviewed (commented):
Looking good. Don't we have sequences to track for generating class, project and batch IDs?

technicallyty commented:
> Looking good. Don't we have sequences to track for generating class, project and batch IDs?

right, so these would be separate messages/tables right? i think batches are covered now with the addition of the auto_increment id

aaronc commented:
> 

Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?

aaronc commented:
We also need balances and supply

technicallyty commented:
> Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?

actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?

i've added a project seq table, and credit/basket balance. looks like the supply is already tracked in `BatchInfo` via `total_amount`.

aaronc commented:
> > Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?
> 
> actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?
> 
> i've added a project seq table, and credit/basket balance. looks like the supply was already tracked in `BatchInfo` via `total_amount`.

I sort of have the preference that things that get updated somewhat frequently - like sequences and supply - should live on separate tables from things which rarely or never change. It's not a huge deal, but I want to avoid the pattern that is done some places in the SDK like x/auth where every acct seq update rewrites the whole account with vesting, pub key and all

technicallyty commented:
> > > Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?
> > 
> > 
> > actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?
> > i've added a project seq table, and credit/basket balance. looks like the supply was already tracked in `BatchInfo` via `total_amount`.
> 
> I sort of have the preference that things that get updated somewhat frequently - like sequences and supply - should live on separate tables from things which rarely or never change. It's not a huge deal, but I want to avoid the pattern that is done some places in the SDK like x/auth where every acct seq update rewrites the whole account with vesting, pub key and all

ok, added batch sequence and separate batch supply table

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,225 @@
+}
+
+// CreditTypeSequence stores and increments the sequence number for credit types
+message CreditTypeSequence {
```
What are these sequences being used for?

technicallyty replied:
i believe it was added in the case we want multiple types of "carbon" credits with different parameters, i.e carbon but measured in a different unit. in which case the abbreviation would still contain "C" and then a sequence for each additional type C0, C01, C02 etc

aaronc replied:
that doesn't make sense 🤔 . I think we should nix it. the abbreviation is the primary key and this is controlled by governance

technicallyty replied:
would it still be auto-generated then?

aaronc replied:
credit type was never meant to be auto-generated. it is chosen by governance. i'm confused by where this seq came from tbh

aaronc replied:
no, the abbreviation is chosen by governance. just `C` for instance. let's fix it

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/commits/a40947c8ec4f81850b3491bb44aab356be08f474

technicallyty replied:
added an issue to remove the sequencing logic from the ecocredit module https://github.com/regen-network/regen-ledger/issues/707

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,225 @@
+
+    string project_id = 1;
+
+    uint64 id = 2;
```
```suggestion
    uint64 next_id = 2;
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,225 @@
+    // batch_denom is the denom of the credit batch
+    string batch_denom = 2;
+
+    message CreditBalance {
+        string tradable = 1;
+        string retired = 2;
+    }
+
+    // balance is the balance of both tradable and retired credits
+    CreditBalance balance = 3;
```
```suggestion
        string tradable = 1;
        string retired = 2;
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,225 @@
+}
+
+// BasketBalance stores the amount of credits from a batch in a basket
+message BasketBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
+        primary_key: {fields: "basket_denom,batch_denom"}
+    };
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
+
+    // balance is the amount of ecocredits held in the basket
+    string balance = 3;
+}
```
Let's put this in a basket sub-package

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,220 @@
+    };
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
```
can we reduce storage space using uint64 id's?

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,208 @@
+    // Some of the issued credits may be cancelled and will be removed from
+    // total_amount and tracked in amount_cancelled. total_amount and
+    // amount_cancelled will always sum to the original credit issuance amount.
+    string total_amount = 2;
```
Where are we tracking retired amount? That's different from cancelled

technicallyty replied:
it wasn't tracked in the original design. i can split the supply into tradable and retired

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/commits/40bbe1ed417b578073d2f232230562cd54a61cf5

aaronc commented on proto/regen/ecocredit/v1beta1/baskets/state.proto:
```diff
@@ -0,0 +1,22 @@
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
```
```suggestion
    uint64 batch_id = 2;
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,199 @@
+        index: {id: 2 fields: "start_date"}
+    };
+
+    // id is a an incrementing integer to identify the batch
```
```suggestion
    // id is an auto-incrementing integer to succintly identify the batch
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,199 @@
+    };
+
+    // address is the address of the credit holder
+    string address = 1;
```
let's use `bytes`

technicallyty replied:
should we be using cosmos scalars? or just change to bytes for now

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,199 @@
+    string retired = 4;
+}
+
+// BatchSequence tracks the sequence number for batches within a class
```
```suggestion
// BatchSequence tracks the sequence number for batches within a project
```

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,199 @@
+    };
+
+    // class_id is the id of the class for a batch
+    uint64 class_id = 1;
```
```suggestion
    string project_id = 1;
```

aaronc reviewed (commented):
getting close! thanks for your patience!

aaronc commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -0,0 +1,199 @@
+message EcocreditBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 8,
+        primary_key: {fields: "address,batch_id"}
```
```suggestion
        primary_key: {fields: "address,batch_id"}
        index: {id: 1, fields: "batch_id,address"}
```

aaronc reviewed (approved):
LGTM. Thanks @technicallyty 🎉 !

aaronc commented:
Just need to fix conflicts

ryanchristo commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,98 @@
+// SellOrder represents the information for a sell order.
+message SellOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "order_id"}
+        index: {id: 1 fields: "batch_denom"}
+        index: {id: 2 fields: "owner"}
+    };
```
Following #671, we need to include the `expiration` message field and add an expiration index.

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

ryanchristo replied:
Looks like the expiration index is still missing..

technicallyty replied:
oh whoops, think i missed the index part and only added it as a field 😅 its in there now!

ryanchristo commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,98 @@
+// BuyOrder represents the information for a buy order.
+message BuyOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "buy_order_id"}
+        index: {id: 1 fields: "buyer"}
+    };
```
Following #671, we need to include the expiration message field and add an expiration index.

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

ryanchristo replied:
Same here..

ryanchristo commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,98 @@
+    // disable_partial_fill disables the default behavior of partially filling
+    // buy orders if the requested quantity is not available.
+    bool disable_partial_fill = 7;
+}
```
We need to add `AllowAskDenom` here as well.

technicallyty replied:
https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

ryanchristo commented on proto/regen/ecocredit/v1beta1/baskets/state.proto:
```diff
@@ -0,0 +1,22 @@
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_id is the denom of the batch
+    uint64 batch_id = 2;
```
```suggestion
    // batch_id is the id of the credit batch
    uint64 batch_id = 2;
```

clevinson commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
+// BatchSequence tracks the sequence number for batches within a project
+message BatchSequence {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 10,
```
```suggestion
        id: 9,
```

clevinson commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
+// BatchSupply tracks the supply of a credit batch
+message BatchSupply {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 11,
```
```suggestion
        id: 10,
```

clevinson commented on proto/regen/ecocredit/v1beta1/baskets/state.proto:
```diff
@@ -0,0 +1,22 @@
+// BasketBalance stores the amount of credits from a batch in a basket
+message BasketBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
```
```suggestion
        id: 1,
```

clevinson replied:
Should this be `1`? As its the only table in this file

technicallyty commented on proto/regen/ecocredit/v1beta1/marketplace/state.proto:
```diff
@@ -0,0 +1,129 @@
+            // sell_order_id is the sell order ID against which the buyer is trying to buy.
+            // When sell_order_id is set, this is known as a direct buy order because it
+            // is placed directly against a specific sell order.
+            uint64 sell_order_id = 1;
```
i think that would be out of the scope of this PR. this structure was copied from existing proto defs. so i believe this concern would warrant its own issue

aaronc replied:
why a list?

robert-zaremba replied:
@aaronc to select multiple orders to fill.

aaronc replied:
this model is just based on buying from a specific sell order. it's not a batch operation. it's like a shopping cart where you buy one product from one vendor

technicallyty commented on proto/regen/ecocredit/v1beta1/state.proto:
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
+
+    // next_id is the sequence number for the project
+    uint64 next_id = 2;
+}
```
so the format of the project id is classID + sequence number. i think actually the project_id field should be class_id. but in any case, the 2 fields are needed

aaronc replied:
it should be class_id

robert-zaremba replied:
ah, so `ProjectSequence` is a type for a key.

robert-zaremba replied:
@technicallyty if you will be updating the `project_id` field, let's update the `message` documentation as well.
